cpp/src/arrow/array.h

```diff
@@ -522,6 +522,44 @@ class ARROW_EXPORT ListArray : public Array {
   std::shared_ptr<Array> values_;
 };
 
+// ----------------------------------------------------------------------
+// FixedSizeListArray
+
+/// Concrete Array class for list data
+class ARROW_EXPORT FixedSizeListArray : public Array {
+ public:
+  using TypeClass = FixedSizeListType;
+
+  explicit FixedSizeListArray(const std::shared_ptr<ArrayData>& data);
+
+  FixedSizeListArray(const std::shared_ptr<DataType>& type, int64_t length,
+                     const std::shared_ptr<Buffer>& value_offsets,
+                     const std::shared_ptr<Array>& values,
+                     const std::shared_ptr<Buffer>& null_bitmap = NULLPTR,
+                     int64_t null_count = kUnknownNullCount, int64_t offset = 0);
+
+  const FixedSizeListType* list_type() const;
+
+  /// \brief Return array object containing the list's values
+  std::shared_ptr<Array> values() const;
+
+  std::shared_ptr<DataType> value_type() const;
+
+  // Neither of these functions will perform boundschecking
+  int32_t value_offset(int64_t i) const {
+    i += data_->offset;
+    return static_cast<int32_t>(list_size_ * i);
+  }
+  int32_t value_length(...) const { return list_size_; }
```

Comment:
I think I prefer the explicit `int64_t i` as parameter. A bit better for the ABI.

Reply:
Even though it's independent of index?

Reply:
I would rather remove the argument. Or you can provide two separate overrides. A variadic function looks bizarre here.

Reply:
Allowing it to be called with an argument means I don't have to rewrite pretty-print :)